PEP 756: PyUnicode_Export()

vstinner opened the pull request:
<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--3960.org.readthedocs.build/pep-0756/

<!-- readthedocs-preview pep-previews end -->

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+  view.
+* ``PyUnicode_Import()``: import a Python str object.
+
+In general, ``PyUnicode_Export()`` has a O(1) complexity: no memory
```
```suggestion
In general, ``PyUnicode_Export()`` has an *O*\ (1) complexity: no memory
```
https://devguide.python.org/documentation/markup/#typographic-conventions

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+:pep:`393` "Flexible String Representation" changed string internals in
+Python 3.3 to use three formats:
+
+* UCS1: Unicode range [U+0000; U+00ff], 1 byte per character
+* UCS2: Unicode range [U+0000; U+ffff], 2 bytes per character
+* UCS4: Unicode range [U+0000; U+10ffff], 4 bytes per character
```
Like movies, I think the first one was unnumbered? The others are hyphented:

https://en.wikipedia.org/wiki/Universal_Coded_Character_Set

```suggestion
* UCS: Unicode range [U+0000; U+00ff], 1 byte per character
* UCS-2: Unicode range [U+0000; U+ffff], 2 bytes per character
* UCS-4: Unicode range [U+0000; U+10ffff], 4 bytes per character
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+* UCS4: Unicode range [U+0000; U+10ffff], 4 bytes per character
+
+A Python ``str`` object must always use the most compact format. For
+example, a string which only contain ASCII characters must use the UCS1
+format.
```
```suggestion
example, a string which only contains ASCII characters must use the original
UCS format.
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+The ``PyUnicode_KIND()`` function can be used to know the format used by
+a string:
+
+* ``PyUnicode_1BYTE_KIND``: UCS1, 1 byte per character
+* ``PyUnicode_2BYTE_KIND``: UCS2, 2 bytes per character
+* ``PyUnicode_4BYTE_KIND``: UCS4, 4 bytes per character
```
```suggestion
* ``PyUnicode_1BYTE_KIND``: UCS, 1 byte per character
* ``PyUnicode_2BYTE_KIND``: UCS-2, 2 bytes per character
* ``PyUnicode_4BYTE_KIND``: UCS-4, 4 bytes per character
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+
+Then, one of the following function can be used to access data:
+
+* UCS1: ``PyUnicode_1BYTE_DATA()``
+* UCS2: ``PyUnicode_2BYTE_DATA()``
+* UCS4: ``PyUnicode_4BYTE_DATA()``
```
```suggestion
* UCS: ``PyUnicode_1BYTE_DATA()``
* UCS-2: ``PyUnicode_2BYTE_DATA()``
* UCS-4: ``PyUnicode_4BYTE_DATA()``
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+* ``PyUnicode_2BYTE_KIND``: UCS2, 2 bytes per character
+* ``PyUnicode_4BYTE_KIND``: UCS4, 4 bytes per character
+
+Then, one of the following function can be used to access data:
```
```suggestion
Then, one of the following functions can be used to access data:
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+* UCS2: ``PyUnicode_2BYTE_DATA()``
+* UCS4: ``PyUnicode_4BYTE_DATA()``
+
+To get best performance, a C extension should have 3 code paths for each
```
```suggestion
To get the best performance, a C extension should have 3 code paths for each
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+Limited C API
+-------------
+
+PEP 393 functions such as ``PyUnicode_KIND()`` and
```
```suggestion
:pep:`393` functions such as ``PyUnicode_KIND()`` and
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+the limited C API can only use less efficient code paths and string
+formats.
+
+For example, the Markupsafe project has a C extension specialized for
```
```suggestion
For example, the MarkupSafe project has a C extension specialized for
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+===================================  ========  ===========================
+Constant Identifier                  Value     Description
+===================================  ========  ===========================
+``PyUnicode_FORMAT_UCS1``            ``0x01``  UCS-1 string (``Py_UCS1*``)
```
```suggestion
``PyUnicode_FORMAT_UCS1``            ``0x01``  UCS string (``Py_UCS1*``)
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+Export complexity
+-----------------
+
+In general, an export has a complexity of O(1): no memory copy is
+needed. There are cases when a copy is needed, O(n) complexity:
```
```suggestion
In general, an export has a complexity of *O*\ (1): no memory copy is
needed. There are cases when a copy is needed, *O*\ (*n*) complexity:
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+In general, an export has a complexity of O(1): no memory copy is
+needed. There are cases when a copy is needed, O(n) complexity:
+
+* If UCS2 is requested and the native format is UCS1.
+* If UCS4 is requested and the native format is UCS1 or UCS2.
+* If UTF8 is requested: the string is encoded to UTF-8 at the first
```
```suggestion
* If UCS-2 is requested and the native format is UCS.
* If UCS-4 is requested and the native format is UCS or UCS-2.
* If UTF-8 is requested: the string is encoded to UTF-8 at the first
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+* If UTF8 is requested: the string is encoded to UTF-8 at the first
+  call, and then the encoded UTF-8 string is cached.
+
+To have a O(1) complexity on CPython and PyPy, it's recommended to
```
```suggestion
To have an *O*\ (1) complexity on CPython and PyPy, it's recommended to
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+------------
+
+When the ``PyUnicode_FORMAT_ASCII`` format is request for export, the
+``PyUnicode_FORMAT_UCS1`` export format is used for ASCII and Latin1
```
```suggestion
``PyUnicode_FORMAT_UCS1`` export format is used for ASCII and Latin-1
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+Usage of PEP 393 C APIs
+=======================
+
+A code search on PyPI top 7,500 projects (on March 2024) shows that
```
```suggestion
A code search on PyPI top 7,500 projects (in March 2024) shows that
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+string. It can lead to bugs, or even security vulnerabilities.
+
+Rejecting embedded NUL characters require to scan the string which has
+a O(n) complexity.
```
```suggestion
an *O*\ (*n*) complexity.
```

hugovk commented on peps/pep-0756.rst:
```diff
@@ -0,0 +1,359 @@
+``str`` object can contain arbitrary lone surrogate characters. Example:
+``"\uDC80"``.
+
+Rejecting surogate characters prevents exporting a string which contain
+such character. It can be surprising and annoying since the
+``PyUnicode_Export()`` caller doesn't control the string content.
```
```suggestion
Rejecting surrogate characters prevents exporting a string which contains
such a character. It can be surprising and annoying since the
``PyUnicode_Export()`` caller doesn't control the string contents.
```

hugovk commented:
And remember the CODEOWNERS entry (and for https://github.com/python/peps/pull/3961).

vstinner commented:
@hugovk:
> And remember the CODEOWNERS entry (and for https://github.com/python/peps/pull/3961).

Oops, done.

I addressed most of your remarks.

About UCS-1, it doesn't exist in https://en.wikipedia.org/wiki/Universal_Coded_Character_Set. It's made up in Python and exists as the C type `Py_UCS1`. But I added hyphens to UCS names (UCS-1, UCS-2, UCS-4).
